File: elixir/lib/valid_parentheses.ex
defmodule ValidParentheses do
  @moduledoc """
  See: <https://leetcode.com/problems/valid-parentheses>
  """

  @spec is_valid(s :: String.t()) :: boolean
  def is_valid(s) do
    is_valid(s, [])
  end

  defp is_valid("(" <> string, stack), do: is_valid(string, ["(" | stack])
  defp is_valid("[" <> string, stack), do: is_valid(string, ["[" | stack])
  defp is_valid("{" <> string, stack), do: is_valid(string, ["{" | stack])
  defp is_valid(")" <> string, ["(" | stack]), do: is_valid(string, stack)
  defp is_valid("]" <> string, ["[" | stack]), do: is_valid(string, stack)
  defp is_valid("}" <> string, ["{" | stack]), do: is_valid(string, stack)
  defp is_valid("", []), do: true
  defp is_valid(_, _), do: false
end

File: javascript/src/valid_parentheses.js
const parenMap = new Map([
    ["(", ")"],
    ["[", "]"],
    ["{", "}"],
]);

/**
 * @see <https://leetcode.com/problems/valid-parentheses>
 *
 * @param {string} string
 * @return {boolean}
 */
function isValid(string) {
    const stack = [];
    for (const char of string) {
        if (parenMap.has(char)) {
            stack.push(parenMap.get(char));
        } else if (stack.length !== 0 && stack.at(-1) === char) {
            stack.pop();
        } else {
            return false;
        }
    }
    return stack.length === 0;
}

export {isValid};
